From f45debfb8b0859c013d3793c51988482e62c38e5 Mon Sep 17 00:00:00 2001
From: Huacai Chen <chenhc@lemote.com>
Date: Sun, 28 May 2017 09:29:33 +0800
Subject: [PATCH 321/374] AOSCOS: 8250_pci: Add a new PLX9050 serial port card
 support

[Mingcong Bai: Resolved merge conflicts in
drivers/tty/serial/8250/8250_pci.c, applied parts of the patch to
drivers/tty/serial/8250/8250_pcilib.c per commit 0348386dab37 ("serial:
8250_pci: Add serial8250_pci_setup_port definition in 8250_pcilib.c").]

Signed-off-by: Huacai Chen <chenhc@lemote.com>
Signed-off-by: Mingcong Bai <jeffbai@aosc.io>

Signed-off-by: Kexy Biscuit <kexybiscuit@aosc.io>
Signed-off-by: Mingcong Bai <jeffbai@aosc.io>
---
 drivers/tty/serial/8250/8250_pci.c    | 19 +++++++++++++++++++
 drivers/tty/serial/8250/8250_pcilib.c |  6 ++++++
 2 files changed, 25 insertions(+)

diff --git a/drivers/tty/serial/8250/8250_pci.c b/drivers/tty/serial/8250/8250_pci.c
index a5cd4c6166e6b..85d7cc9683d3f 100644
--- a/drivers/tty/serial/8250/8250_pci.c
+++ b/drivers/tty/serial/8250/8250_pci.c
@@ -349,6 +349,12 @@ static int pci_plx9050_init(struct pci_dev *dev)
 		 * deep FIFOs
 		 */
 		irq_config = 0x5b;
+
+	if ((dev->vendor == PCI_VENDOR_ID_PLX) &&
+	    (dev->device == PCI_DEVICE_ID_PLX_9050) &&
+	    (dev->subsystem_vendor == PCI_VENDOR_ID_PLX) &&
+	    (dev->subsystem_device == PCI_DEVICE_ID_PLX_9050))
+		irq_config = 0x53;
 	/*
 	 * enable/disable interrupts
 	 */
@@ -2405,6 +2411,15 @@ static struct pci_serial_quirk pci_serial_quirks[] = {
 		.setup		= pci_default_setup,
 		.exit		= pci_plx9050_exit,
 	},
+	{
+		.vendor		= PCI_VENDOR_ID_PLX,
+		.device		= PCI_DEVICE_ID_PLX_9050,
+		.subvendor	= PCI_VENDOR_ID_PLX,
+		.subdevice	= PCI_DEVICE_ID_PLX_9050,
+		.init		= pci_plx9050_init,
+		.setup		= pci_default_setup,
+		.exit		= pci_plx9050_exit,
+	},
 	{
 		.vendor		= PCI_VENDOR_ID_PLX,
 		.device		= PCI_DEVICE_ID_PLX_9050,
@@ -4543,6 +4558,10 @@ static const struct pci_device_id serial_pci_tbl[] = {
 		PCI_SUBDEVICE_ID_UNKNOWN_0x1584, 0, 0,
 		pbn_b2_4_115200 },
 	/* Unknown card - subdevice 0x1588 */
+	{	PCI_VENDOR_ID_PLX, PCI_DEVICE_ID_PLX_9050,
+		PCI_VENDOR_ID_PLX,
+		PCI_DEVICE_ID_PLX_9050, 0, 0,
+		pbn_b2_8_921600 },
 	{	PCI_VENDOR_ID_PLX, PCI_DEVICE_ID_PLX_9050,
 		PCI_VENDOR_ID_PLX,
 		PCI_SUBDEVICE_ID_UNKNOWN_0x1588, 0, 0,
diff --git a/drivers/tty/serial/8250/8250_pcilib.c b/drivers/tty/serial/8250/8250_pcilib.c
index 9d5d2531a33bd..66e21ce0a9974 100644
--- a/drivers/tty/serial/8250/8250_pcilib.c
+++ b/drivers/tty/serial/8250/8250_pcilib.c
@@ -27,6 +27,12 @@ int serial8250_pci_setup_port(struct pci_dev *dev, struct uart_8250_port *port,
 	if (bar >= PCI_STD_NUM_BARS)
 		return -EINVAL;
 
+	if ((dev->vendor == PCI_VENDOR_ID_PLX) &&
+	    (dev->device == PCI_DEVICE_ID_PLX_9050) &&
+	    (dev->subsystem_vendor == PCI_VENDOR_ID_PLX) &&
+	    (dev->subsystem_device == PCI_DEVICE_ID_PLX_9050))
+		offset += 0x80;
+
 	if (pci_resource_flags(dev, bar) & IORESOURCE_MEM) {
 		port->port.iotype = UPIO_MEM;
 		port->port.iobase = 0;
-- 
2.52.0

